File: pathfinding/graph-traversals/python/graph_bfs.py
# Raspberry Pi Foundation
# Developed to be used alongside Isaac Computer Science,
# part of the National Centre for Computing Education
# Usage licensed under CC BY-SA 4


def breadth_first_search(graph, start_node, target_node):
    """A breadth-first search performed on a graph stored as a dictionary"""

    # Initialisation
    queue = [start_node]
    discovered = [start_node]
    neighbours = []
    found = False

    # Repeat while there are items in the queue
    # and the target node has not been found 
    while len(queue) != 0 and found == False:
        
        # Set the current node to the first item in the queue
        current_node = queue[0]

        # Remove the current node from the start of the queue
        del queue[0]

        # Get the current node's list of neighbours
        neighbours = graph[current_node]

        # Testing
        print(f"\nQueue: {queue}")
        print(f"Discovered: {discovered}")
        print(f"Neighbours: {neighbours}")

        # Repeat for each node in the neighbours list
        for node in neighbours:
            # Check if the node has not already been discovered
            if node not in discovered:
                # Check if the target node has been found
                if node == target_node:
                    found = True
                else:
                    # Add the node to the stack
                    # and to the discovered list
                    queue.append(node)
                    discovered.append(node)

                    # Testing
                    print(f"\nQueue: {queue}")
                    print(f"Discovered: {discovered}")
                    print(f"Neighbours: {neighbours}")
                    
    return found




def main():
    """Create a graph and search for an item"""

    # Use a dictionary to represent the graph as an adjacency list.
    # Each key is a node in the graph.
    # Each value is a list of the node's neighbours
    test_graph = {
        "1": ["2", "9"],
        "2": ["1"],
        "3": ["4", "5", "6", "9"],
        "4": ["3"],
        "5": ["3", "8"],
        "6": ["3", "7"],
        "7": ["6", "8", "9"],
        "8": ["5", "7"],
        "9": ["1", "3", "7"]
        }

    print("### Graph traversal - breadth-first search (BFS) ###")

    # Search for a value and return True if it has been found
    item_to_find = "4"
    start = "1"
    found = breadth_first_search(test_graph, start, item_to_find)

    # Output the search result
    print(f"\nThe target node is {item_to_find}")
    
    if found == True:
        print(f"{item_to_find} was found in the graph")
    else:
        print(f"{item_to_find} was NOT found in the graph")


# This code will run if this file is executed directly
# (i.e. not called by another program)
if __name__ == '__main__':
    main()

File: pathfinding/graph-traversals/c-sharp/graph_bfs.cs
/*
Isaac Computer Science
Usage licensed under the Open Government Licence v3.0

Note: This file is designed to be copied out and compiled on your machine.
In order for it to compile properly you need to ensure that the project name is the same as the "namespace" in this file.
To run this file you need to:
1. Copy the contents
2. Paste them into the C# IDE of your choice (Visual Studio, for example)
3. Change the namespace to match your project (if necessary)
4. Compile the program
5. Run the program
*/

using System;
using System.Collections.Generic;

namespace IsaacCodeSamples
{
    class Graph
    {
        // The Main method is the entry point for all C# programs
        public static void Main()
        {
            // Represent a graph using a dictionary with lists as values
            var testGraph = new Dictionary<string, List<string>>() {
                {"1", new List<string>() {"2", "9"} },
                {"2", new List<string>() {"1"} },
                {"3", new List<string>() {"4", "5", "6", "9"} },
                {"4", new List<string>() {"3"} },
                {"5", new List<string>() {"3", "8"} },
                {"6", new List<string>() {"3", "7"} },
                {"7", new List<string>() {"6", "8"} },
                {"8", new List<string>() {"5", "7"} },
                {"9", new List<string>() {"1", "3", "7"} }
            };

            Console.WriteLine("### Graph traversal - breadth-first search (BFS) ###");

            // Search for a value and return if it has been found
            string itemToFind = "4";
            string start = "1";
            bool found = BreadthFirstSearch(testGraph, start, itemToFind);

            // Output the search result
            Console.WriteLine($"\nThe target node is {itemToFind}");
            if (found == true) {
                Console.WriteLine($"{itemToFind} was found in the graph");
            }
            else {
                Console.WriteLine($"{itemToFind} does not exist in the graph");
            }
        }

        // A breadth-first search performed on a graph
        public static bool BreadthFirstSearch(Dictionary<string, List<string>> graph, 
            string startNode, string targetNode)
        {
            // Initialise the variables
            List<string> queue = new List<string>() {startNode};
            List<string> discovered = new List<string>() {startNode};
            List<string> neighbours = new List<string>();
            bool found = false;

            // Repeat while there are items in the queue
            // and the target node has not been found 
            while (queue.Count != 0 && found == false) {
                
                // Set the current node to the first item in the queue
                string currentNode = queue[0];

                // Remove the current node from the start of the queue
                queue.RemoveAt(0);

                // Get the current node's list of neighbours
                neighbours = graph[currentNode];

                // Testing
                Console.Write("\nQueue: ");
                Console.WriteLine("[{0}]", string.Join(", ", queue));
                Console.Write("Discovered: ");
                Console.WriteLine("[{0}]", string.Join(", ", discovered));
                Console.Write("Neighbours: ");
                Console.WriteLine("[{0}]", string.Join(", ", neighbours));

                // Repeat for each node in the neighbours list
                foreach (string node in neighbours) {
                    // Check if the node has not already been discovered
                    if (discovered.Contains(node) == false) {
                        // Check if the target node has been found
                        if (node == targetNode) {
                            found = true;
                        }
                        // Otherwise append the node to the queue and discovered lists
                        else {
                            queue.Add(node);
                            discovered.Add(node);

                            // Testing
                            Console.Write("\nQueue: ");
                            Console.WriteLine("[{0}]", string.Join(", ", queue));
                            Console.Write("Discovered: ");
                            Console.WriteLine("[{0}]", string.Join(", ", discovered));
                            Console.Write("Neighbours: ");
                            Console.WriteLine("[{0}]", string.Join(", ", neighbours));
                        }
                    }
                }
            }
            return found;
        }

        
    }
}